text/0000-buildpack-registry-search-api.md

```diff
@@ -0,0 +1,235 @@
+# Meta
+[meta]: #meta
+- Name: Buildpack Registry Search API
+- Start Date: 2020-11-11
+- Author(s): @elbandito
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+This is a proposal for a service API that enables clients to search for buildpacks.
+
+# Definitions
+[definitions]: #definitions
+
+**API** - Application Programmer Interface
+
+**Buildpack Registry** - the CNB centeralized [registry](https://github.com/buildpacks/registry-index) that contains metadata for all the community supported buildpacks 
+
+**Service** - an independent web process running in the cloud that provides RESTful endpoints
+
+**Client** - represents a physical end-user or software interface e.g. CLI, web UI, etc.
+ 
+# Motivation
+[motivation]: #motivation
+
+As the CNB project makes progress towards supporting a centralized Buildpack Registry, I anticipate the need to quickly search for and identify specific buildpacks.  To support this, I propose the creation of an API search service that customers could use directly and/or indirectly via a web dashboard or CLI.
+
+- What use cases does it support?  
+  - Customers looking to quickly indentify usefulful buildpacks to build their applications.
+- What is the expected outcome?  
+  - Customers will be able to quickly find relevant buildpacks for their use cases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+An external API service that exposes endpoints for retrieving buildpack metadata.  Initially, I envision this service providing a few GET endpoints to support plain text query searches, and individual buildpack data retrieval.  The API endpoints will be versioned, and follow OpenAPI and/or Json Schema/API standards.
+
+### Versioned Endpoints
+
+All the endpoints defined below will be version via a `vN` prefix to the base URL. e.g. `https://registry.buildpacks.io/v1/<endpoint>`.
+
+### Supported Endpoints
+
+- **GET /search?matches=text**
+
+  Retrieves all the buildpacks that satisfy the search query.  If the response contains more than significant number of buildpacks, pagination will be used.  The current approach to pagination is to take inspiration from how [Github](https://docs.github.com/en/free-pro-team@latest/rest/guides/traversing-with-pagination).  
+  ```
+  $ curl https://registry.buildpacks.io/v1/search?matches=projectriff
```

Comment:
do we want to namespace this with "api", i.e. <https://registry.buildpacks.io/v1/search?matches=projectriff>? This way <https://registry.buildpacks.io> could be where the frontend lives.

Reply:
@hone like: `api.registry.buildpacks.io` or `api.buildpacks.io`?

Reply:
FWIW: From my systems management days, I like subdomain based routing plus it looks nice. +1 on `https://api.buildpacks.io`.

Reply:
This isn't really intended to be an api for all of buildpack.io though. I think it should be:

* `https://registry.buildpacks.io` - frontend
* `https://registry.buildpacks.io/api` - API

Reply:
Yeah, I didn't want to presume this would be an API for all of buildpacks or what that means. At that point it'd be <https://api.buildpacks.io/registry>?